presto-spi/src/main/java/com/facebook/presto/spi/block/RowBlockEncoding.java

```diff
@@ -40,24 +32,21 @@ public String getName()
     }
 
     @Override
-    public void writeBlock(SliceOutput sliceOutput, Block block)
+    public void writeBlock(BlockEncodingSerde blockEncodingSerde, SliceOutput sliceOutput, Block block)
     {
         AbstractRowBlock rowBlock = (AbstractRowBlock) block;
-
-        if (rowBlock.numFields != fieldBlockEncodings.length) {
-            throw new IllegalArgumentException(
-                    "argument block differs in length (" + rowBlock.numFields + ") with this encoding (" + fieldBlockEncodings.length + ")");
-        }
+        int numFields = rowBlock.numFields;
 
         int positionCount = rowBlock.getPositionCount();
 
         int offsetBase = rowBlock.getOffsetBase();
         int[] fieldBlockOffsets = rowBlock.getFieldBlockOffsets();
-
         int startFieldBlockOffset = fieldBlockOffsets[offsetBase];
         int endFieldBlockOffset = fieldBlockOffsets[offsetBase + positionCount];
-        for (int i = 0; i < fieldBlockEncodings.length; i++) {
-            fieldBlockEncodings[i].writeBlock(sliceOutput, rowBlock.getFieldBlocks()[i].getRegion(startFieldBlockOffset, endFieldBlockOffset - startFieldBlockOffset));
+
+        sliceOutput.appendInt(numFields);
+        for (int i = 0; i < numFields; i++) {
+            blockEncodingSerde.writeBlock(sliceOutput, rowBlock.getFieldBlocks()[i].getRegion(startFieldBlockOffset, endFieldBlockOffset - startFieldBlockOffset));
```

Comment:
same comment about a helper method

Reply:
Done renaming. I will try out introducing the helper separately.